File: leetcode-72.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n = len(word1)
        m = len(word2)
        dp = [[0]*(m+1) for i in range(n+1)]
        if n == 0:
            return m
        if m == 0:
            return n
        for i in range(1, m+1):
            dp[0][i] = i
        for i in range(n+1):
            dp[i][0] = i
        for i in range(1, n+1):
            for j in range(1, m+1):
                if word1[i-1] == word2[j-1]:
                    dp[i][j] = dp[i-1][j-1]
                else:
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1])+1
        return dp[-1][-1]

if __name__ == "__main__":
    so = Solution()
    print(so.minDistance('a', 'a'))

File: leetcode/leetcode-72.cpp
/*
 *@File Name          :leetcode-72.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/3/12 9:10
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "string"
#include "vector"

using namespace std;

class Solution {
public:
    int minDistance(string word1, string word2) {
        int n = word1.size();
        int m = word2.size();
        if (n == 0 || m == 0) return max(n, m);
        vector<vector<int>> dp(n + 1, vector<int>(m + 1, 0));
        for (int i = 1; i <= n; i++)
            dp[i][0] = i;
        for (int i = 1; i <= m; i++)
            dp[0][i] = i;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (word1[i] == word2[j]) {
                    dp[i + 1][j + 1] = dp[i][j];
                } else {
                    dp[i + 1][j + 1] = min(dp[i][j], min(dp[i][j + 1], dp[i + 1][j])) + 1;
                }
            }
        }
        return dp[n][m];
    }
};